packages/preferences-persistence/src/create/create-async-debouncer.js

```diff
@@ -9,23 +9,21 @@
  * This is distinct from `{ debounce } from @wordpress/compose` in that it
  * waits for promise resolution.
  *
- * @param {Function} func    A function that returns a promise.
- * @param {number}   delayMS A delay in milliseconds.
- *
- * @return {Function} A function that debounce whatever function is passed
- *                    to it.
+ * @return {Function} A function that debounces the async function passed to it
+ *                    in the first parameter by the time passed in the second
+ *                    options parameter.
  */
-export default function debounceAsync( func, delayMS ) {
+export default function createAsyncDebouncer() {
 	let timeoutId;
 	let activePromise;
 
-	return async function debounced( ...args ) {
+	return async function debounced( func, { delayMS, isTrailing = false } ) {
 		// This is a leading edge debounce. If there's no promise or timeout
 		// in progress, call the debounced function immediately.
-		if ( ! activePromise && ! timeoutId ) {
+		if ( ! isTrailing && ! activePromise && ! timeoutId ) {
```

Comment:
I think this debounce logic has been a little bit flawed from the beginning, and the long `expensiveRequestDebounceMS` makes it worse.

If I call the debounced function repeatedly, the timeout keeps getting cancelled and rescheduled, and as a result the save can be never performed. Or only after a very long time, much longer than the intended timeout.

To avoid this "starvation" problem, we should do throttling rather than debouncing. Once the 60sec timeout is active, keep it running, but on further calls only update the _function_ to be called when it finally expires.

The Lodash [`debounce`](https://lodash.com/docs/4.17.15#debounce) function has a `maxWait` option that achieves something similar.

If we fix this behavior, then maybe we can remove also the invocation on the leading edge? I don't know why it was originally added -- why can't we always wait 2.5s? Maybe it's a workaround for this starvation problem?

Reply:
Just noting that we already have both `throttle` and `debounce` utilities as part of the `@wordpress/compose` package, and the built-in `debounce()` also supports a `maxWait` option. Just in case that could be something you'd like to utilize instead of a custom debounce logic.

Reply:
> If I call the debounced function repeatedly, the timeout keeps getting cancelled and rescheduled, and as a result the save can be never performed. Or only after a very long time, much longer than the intended timeout.
>To avoid this "starvation" problem, we should do throttling rather than debouncing. Once the 60sec timeout is active, keep it running, but on further calls only update the function to be called when it finally expires.

I'll look into it. It's not really much of an issue with the shorter delay, but perhaps becomes more apparent with the longer delay. 🤔 

The main aim is to avoid repeated REST requests. WordPress servers are very bad at handling multiple requests at the same time, but perhaps it's ok to throttle.

> If we fix this behavior, then maybe we can remove also the invocation on the leading edge? I don't know why it was originally added -- why can't we always wait 2.5s? Maybe it's a workaround for this starvation problem?

The most common use case is infrequent preferences calls, so triggering the request right away is usually fine. It also reduces the chances that the delay/request might be in flight while the user unloads the page, causing the request to never go through.

> Just noting that we already have both throttle and debounce utilities as part of the @wordpress/compose package, and the built-in debounce() also supports a maxWait option. Just in case that could be something you'd like to utilize instead of a custom debounce logic.

@tyxla It needs to support async functions, which most libraries don't. The code here ensures requests are never made in parallel—it delays from after a promise completion rather than from function invocation. It might be there's another way of doing this that allows leverage of worpdress/compose, any suggestions appreciated.

Reply:
I've updated it to work as @jsnajdr describes so that we can try it out.

I think it works well. I've kept the debounce for the non-expensive call a short 1 second, and the expensive one 5 seconds, and there's a 10 second max wait.

The values feel a bit arbitrary, but I figure it should be ok to prevent spamming of HTTP requests if the user is duplicating lots of blocks.

The function is now also very simple, which is a bonus.

I think all the other feedback has also been addressed. I still need to give this a thorough test as so long has passed since the last time I looked at this PR.